### ISSUE TITLE
Syntax Error for custom property sets in no-descending-specificity

### ISSUE THREAD

**Issue by woble** [stylelint/stylelint#1999]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

When using [postcss-apply](https://github.com/pascalduez/postcss-apply) in combination with `no-descending-specificity` rule, two things happen.

1. Using `gulp-postcss` errors out.
2. Using `CLI` return unexpected output.

> Which rule, if any, is this issue related to?

`no-descending-specificity`

> What CSS is needed to reproduce this issue?

```css
:root {
    --foo: {
        color: white;
        background-color: black;
    };
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
    "rules":
    {
        "no-descending-specificity": true
    }
}
```

> Which version of stylelint are you using?

e.g. `7.4.2`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

1. Using `gulp-stylelint` returns an error and breaks the build process because, presumably, the gulp-stylelint can't handle what stylelint is returning.
```
TypeError in plugin 'gulp-stylelint'
Message:
    Cannot read property 'toUpperCase' of undefined
Details:
    domainEmitter: [object Object]
    domain: [object Object]
    domainThrown: false
```

2. Using `cli` (`stylelint foo.pcss --config stylelint.json`) returns just this with above example:
`2:5     Cannot parse selector`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, use of `postcss-apply`.

> What did you expect to happen?

No warnings to be flagged.

**Comment by evilebottnawi** [comment]
@woble can you show full stack trace?

**Comment by jeddy3** [comment]
This is probably because custom property sets [are parsed as rules](https://astexplorer.net/#/W1mMdWlyqF) in PostCSS.

It might simply be the case of returning early [here](https://github.com/stylelint/stylelint/blob/master/src/rules/no-duplicate-selectors/index.js#L35) if the "rule" is actually a [custom property set](https://github.com/stylelint/stylelint/blob/master/src/utils/isCustomPropertySet.js). 

@woble Would you like to try this?

**Comment by woble** [comment]
That was my assumption as well. Can you provide what I need to test and I will.

@evilebottnawi what full stack of do you want exactly? Everything I posted is all I get for stylelint.

**Comment by evilebottnawi** [comment]
@woble 
```shell
TypeError in plugin 'gulp-stylelint'
Message:
    Cannot read property 'toUpperCase' of undefined
Details:
    domainEmitter: [object Object]
    domain: [object Object]
    domainThrown: false
```

**Comment by evilebottnawi** [comment]
@woble full stacktrace of error with error line and column and stack of functions calls

**Comment by jeddy3** [comment]
> Can you provide what I need to test and I will.

To paraphrase [the developer guide](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-bugs):

> Write failing unit tests that exemplify the bug:

Adding the following [here](https://github.com/stylelint/stylelint/blob/master/src/rules/no-descending-specificity/__tests__/index.js#L59) should be enough:

```css
:root { --foo: {}; }
```

> Fiddle with the rule until those new [tests pass](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#running-tests).

Use the following to run tests:

```shell
./node_modules/.bin/babel-tape-runner src/rules/no-descending-specificity/__tests__/index.js
```

Good luck :)

**Comment by woble** [comment]
Thanks, I will see what I can produce.

@evilebottnawi
The stuff I provided was output in a console and it does not appear to expand objects. Any suggestions on how to easily produce it outside the console?

**Comment by evilebottnawi** [comment]
@woble https://github.com/stylelint/stylelint/blob/master/src/standalone.js#L119 -> `console.log(error)`

**Comment by woble** [comment]
PostCSS does indeed treat it as a selector, as @jeddy3 showed. Adding the following [here](https://github.com/stylelint/stylelint/blob/master/src/rules/no-descending-specificity/index.js#L37) solves the problem with both standalone (no `Cannot parse selector` message) and gulp-stylelint (no runtime error).

```js
// Ignore custom property set `--foo: {};`
if (/^--.+:$/.test(trimSelector)) { return }
```

Would that be sufficient enough for you guys?

**Comment by davidtheclark** [comment]
Closed by #2000.
